eigen: support sparse matrix types

hmenke opened the pull request:
This is a first attempt at supporting sparse matrix types from Eigen. The code is mostly ported over from [pybind11](https://github.com/pybind/pybind11/blob/master/include/pybind11/eigen/matrix.h). It compiles but the test segfaults. This could have to with the fact that I have not really a good idea of what I'm doing. Help is very much appreciated.

wjakob commented on include/nanobind/eigen/dense.h:
```diff
@@ -12,9 +12,10 @@
 
 #include <nanobind/tensor.h>
 #include <Eigen/Core>
+#include <Eigen/SparseCore>
```
Let's put the whole sparse-specific parts into a new headerfile `nanobind/eigen/sparse.h` that includes `nanobind/eigen/dense.h`

wjakob replied:
This means that `is_eigen_sparse_v` and `is_eigen_xpr_v` have to be defined in a way that doesn't depend on elements in `Eigen/SparseCore` -- perhaps there is a constexpr field like `IsSparse` that can be used instead (or another field that is only defined for sparse types).

hmenke replied:
I've moved everything to `nanobind/eigen/sparse.h`, but `is_eigen_sparse_v` cannot be moved (yet) unfortunately, because `is_eigen_v` matches `Eigen::MatrixBase` which `Eigen::SparseMatrix` inherits from. There is no member like `IsSparse` by we could maybe try to detect the `valuePtr()`, `outerIndexPtr()`, and `innerIndexPtr()` methods to decide. This would however entail SFINAE which is slow to compile

wjakob replied:
`SparseMatrixBase` has a forward declaration in `Eigen/src/util/ForwardDeclarations.h`. Isn't it possible to do the test without pulling in the whole `SparseCore` header?

hmenke replied:
Yes, this does work indeed. Good find! :ok_hand:

wjakob commented on include/nanobind/eigen/dense.h:
```diff
@@ -241,6 +249,107 @@ struct type_caster<Eigen::Ref<T, Options, StrideType>> {
+/// Caster for Eigen::SparseMatrix<T>
+template <typename T> struct type_caster<T, enable_if_t<is_eigen_sparse_v<T>>> {
+    using Scalar = typename T::Scalar;
+    using StorageIndex = std::remove_reference_t<decltype(*std::declval<T>().outerIndexPtr())>;
```
That definition of `StorageIndex` strikes me as surprisingly complicated. Can we not get the type more directly from a `using` alias defined within `T` (e.g. `T::StorageIndex`?)

wjakob commented on include/nanobind/eigen/dense.h:
```diff
@@ -241,6 +249,107 @@ struct type_caster<Eigen::Ref<T, Options, StrideType>> {
+    IndexCaster index_caster;
+
+    bool from_python(handle src, uint8_t flags, cleanup_list *cleanup) noexcept {
+	if (!src) {
```
those three lines are not needed in nanobind.

wjakob commented on include/nanobind/eigen/dense.h:
```diff
@@ -241,6 +249,107 @@ struct type_caster<Eigen::Ref<T, Options, StrideType>> {
+	object matrix_type = sparse_module.attr(rowMajor ? "csr_matrix" : "csc_matrix");
+
+	if (!obj.type().is(matrix_type)) {
+            try {
```
expand this `try` block so that it also includes the `module::import_`, etc. bits. These can raise exceptions, and the type caster is *not* allowed to raise.

wjakob commented on include/nanobind/eigen/dense.h:
```diff
@@ -241,6 +249,107 @@ struct type_caster<Eigen::Ref<T, Options, StrideType>> {
+    using Scalar = typename T::Scalar;
+    using StorageIndex = std::remove_reference_t<decltype(*std::declval<T>().outerIndexPtr())>;
+    using Index = typename T::Index;
+    static constexpr bool rowMajor = T::IsRowMajor;
```
use `underscore_case`

wjakob commented on include/nanobind/eigen/dense.h:
```diff
@@ -241,6 +249,107 @@ struct type_caster<Eigen::Ref<T, Options, StrideType>> {
+            }
+        }
+
+	if (!scalar_caster.from_python(obj.attr("data"), flags, cleanup)) {
```
I would save the output of `obj.attr("data")` into a `object` variable to ensure that the reference is kept alive while `from_python` runs (you wrote code that does an implicit cast to a `handle` that doesn't hold a reference, and that may cause undefined behavior)

Applies to other casts below as well.

wjakob commented on include/nanobind/eigen/dense.h:
```diff
@@ -241,6 +249,107 @@ struct type_caster<Eigen::Ref<T, Options, StrideType>> {
+
+	auto obj = borrow<object>(src);
+	object sparse_module = module_::import_("scipy.sparse");
+	object matrix_type = sparse_module.attr(rowMajor ? "csr_matrix" : "csc_matrix");
```
You aren't using the `sparse_module` variable elsewhere, so I'd merge those two lines (`module::import_(..).attr(..)`).

wjakob commented on include/nanobind/eigen/dense.h:
```diff
@@ -241,6 +249,107 @@ struct type_caster<Eigen::Ref<T, Options, StrideType>> {
+    using Index = typename T::Index;
+    static constexpr bool rowMajor = T::IsRowMajor;
+
+    using ScalarCaster = make_caster<tensor<Scalar>>;
```
Your specification seems too general (we wouldn't want this to accept a matrix, non-contiguous representations, or GPU arrays). I would write
```cpp
using ScalarTensor = tensor<Scalar, shape<any>, c_contig, cpu>;
```

and then a few lines below:
```
make_caster<ScalarTensor> scalar_caster;
```

wjakob commented on include/nanobind/eigen/dense.h:
```diff
@@ -241,6 +249,107 @@ struct type_caster<Eigen::Ref<T, Options, StrideType>> {
+	if (!storageindex_caster.from_python(obj.attr("indices"), flags, cleanup)) {
+	    return false;
+	}
+        auto innerIndices = storageindex_caster.value;
```
use `undescore_case` for variable names.

I really don't like `auto` when it is avoidable -- can you use a type definition? (`ScalarTensor&`, `IndexTensor&`, etc.)

Both comments apply to a few other places as well.

wjakob commented on include/nanobind/eigen/dense.h:
```diff
@@ -241,6 +249,107 @@ struct type_caster<Eigen::Ref<T, Options, StrideType>> {
+	}
+        auto innerIndices = storageindex_caster.value;
+
+	if (!scalar_caster.from_python(obj.attr("indptr"), flags, cleanup)) {
```
Minor style nit pick: throughout the codebase, short 1-line if statements are written

```cpp
if (x)
    return false;
```

as opposed to

```cpp
if (x) {
    return false;
}
```

wjakob commented on include/nanobind/eigen/dense.h:
```diff
@@ -241,6 +249,107 @@ struct type_caster<Eigen::Ref<T, Options, StrideType>> {
+        auto outerIndices = storageindex_caster.value;
+
+	// FIXME: Write with caster.from_python
+	auto shape = cast<std::tuple<Index, Index>>(obj.attr("shape"));
```
I am surprised that this works, because the `std::tuple` type caster hasn't been imported (and that seems like a rather heavy dependency). We really don't want to pull in any STL type casters since that may lead to unexpected consequences in user code.

Can you do it manually, e.g.:
```cpp
object shape_o = obj.attr("shape"), nnz_o = obj.attr("nnz");
int rows, cols, nnz;
try {
    if (len(shape_o) != 2)
       return false;
    rows = cast<int>(shape_o[0]);
    cols = cast<int>(shape_o[1]);
    nnz = cast<Index>(nnz_o);
} catch (const python_error &) {
    return false;
}
```

wjakob commented on include/nanobind/eigen/dense.h:
```diff
@@ -241,6 +249,107 @@ struct type_caster<Eigen::Ref<T, Options, StrideType>> {
+	}
+        auto nnz = index_caster.value;
+
+        value = EigenMapSparseMatrix<Scalar,
```
I don't remember all the details of the pybind11 version of this code, but a general question: who will keep this `Eigen::Map` alive? Is the idea that this will immediately be cast into a concrete Eigen sparse type that makes a copy?

hmenke replied:
Either assignment to an `Eigen::SparseMatrix` for copy or possibly assignment to another `Eigen::Map<Eigen::SparseMatrix>` for zero-copy (if that works?). If the latter does not work, then using an owning `Eigen::SparseMatrix` here will be better. I also don't know the pybind11 code, I just took “inspiration” from it :)

wjakob commented on include/nanobind/eigen/dense.h:
```diff
@@ -241,6 +249,107 @@ struct type_caster<Eigen::Ref<T, Options, StrideType>> {
+    }
+
+    static handle from_cpp(const T &v, rv_policy policy, cleanup_list *cleanup) noexcept {
+	// FIXME: Looks like UB. Is there anything that can be done about this?
```
did pybind11 use the same code?

hmenke replied:
Yes. https://github.com/pybind/pybind11/blob/master/include/pybind11/eigen/matrix.h#L677

wjakob commented on include/nanobind/eigen/dense.h:
```diff
@@ -241,6 +249,107 @@ struct type_caster<Eigen::Ref<T, Options, StrideType>> {
+	T & src = const_cast<T &>(v);
+	src.makeCompressed();
+
+	object sparse_module = module_::import_("scipy.sparse");
```
merge with line below

wjakob commented on include/nanobind/eigen/dense.h:
```diff
@@ -241,6 +249,107 @@ struct type_caster<Eigen::Ref<T, Options, StrideType>> {
+	object matrix_type = sparse_module.attr(rowMajor ? "csr_matrix" : "csc_matrix");
+
+	// FIXME: These casts are ugly! Is there another way to write this?
+	const size_t data_shape[] = { static_cast<size_t>(v.nonZeros()) };
```
using `static_cast` for an integer conversion seems quite overkill.

wjakob replied:
Is there something else that bothers you here?

hmenke replied:
The cast is needed to narrow the results of the calls to `size_t` which is required for the `nb::tensor` constructor. Leaving out the cast generates a compiler warning in GCC and an error in Clang:
```
nanobind/include/nanobind/eigen/sparse.h:115:32: error: non-constant-expression cannot be narrowed from type 'Eigen::Index' (aka 'long') to 'size_t' (aka 'unsigned long') in initializer list [-Wc++11-narrowing]
        const size_t data_shape[] = { v.nonZeros() };
                                      ^~~~~~~~~~~~
```
Why do you think `static_cast` is overkill? What's the alternative?

wjakob replied:
Just a c-style cast. I think static_cast is great when you have polymorphic classes and other complex aspects of C++ and want to be extra-clear about what kind of cast is acceptable. But I don't see the advantage for simple integer casts, and it just makes the code less readable IMO.

hmenke replied:
But according to https://eel.is/c++draft/expr.cast `(T)` will just try in order `const_cast<T>`, `static_cast<T>`, `reinterpret_cast<T>`. See also https://stackoverflow.com/questions/103512/why-use-static-castintx-instead-of-intx. But I'm happy to do C-style casts as well if you wish.

wjakob replied:
Right. In those stackoverflow threads, you can see that there are footguns involving C++ classes, implicit conversions involving them, and polymorphism. None of which applies to a simple integer cast. For me, using static_cast even in such simple situations is an anti-pattern that makes code ugly to read. (This is somewhat subjective, I know that other people may disagree on what a good default policy is).

wjakob reviewed (commented):
Thanks for the PR! Here is a first bunch of feedback.

hmenke commented:
Thank you very much for the thorough review. I've pushed a fixup commit which I will later rebase, but for the sake of review I will leave it split for now to better follow the progress.

hmenke commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,125 @@
+                               std::move(data), std::move(innerIndices), std::move(outerIndices)),
+                           make_tuple(v.rows(), v.cols()))
+            .release();
+    }
```
Currently the test segfaults in the return statement, but I have no idea where this comes from.

<details>

<summary>Backtrace</summary>

```cpp
Thread 1 (Thread 0x7ffff7891740 (LWP 143127) "python"):
#0  0x00007ffff791ebc7 in __pthread_kill_implementation () from /nix/store/9xfad3b5z4y00mzmk2wnn4900q0qmxns-glibc-2.35-224/lib/libc.so.6
No symbol table info available.
#1  0x00007ffff78d1b46 in raise () from /nix/store/9xfad3b5z4y00mzmk2wnn4900q0qmxns-glibc-2.35-224/lib/libc.so.6
No symbol table info available.
#2  0x00007ffff78bc4b5 in abort () from /nix/store/9xfad3b5z4y00mzmk2wnn4900q0qmxns-glibc-2.35-224/lib/libc.so.6
No symbol table info available.
#3  0x00007fffae4899fd in __gnu_cxx::__verbose_terminate_handler() [clone .cold] () from /home/henri/Code/nanobind/build/tests/test_eigen_ext.cpython-310-x86_64-linux-gnu.so
No symbol table info available.
#4  0x00007fffae4ffdca in __cxxabiv1::__terminate(void (*)()) () from /home/henri/Code/nanobind/build/tests/test_eigen_ext.cpython-310-x86_64-linux-gnu.so
No symbol table info available.
#5  0x00007fffae4ffe35 in std::terminate() () from /home/henri/Code/nanobind/build/tests/test_eigen_ext.cpython-310-x86_64-linux-gnu.so
No symbol table info available.
#6  0x00007fffae49633f in __clang_call_terminate () from /home/henri/Code/nanobind/build/tests/test_eigen_ext.cpython-310-x86_64-linux-gnu.so
No symbol table info available.
#7  0x00007fffae4b3878 in nanobind::detail::type_caster<Eigen::SparseMatrix<float, 1, int>, int>::from_cpp (v=..., policy=nanobind::rv_policy::move, cleanup=0x7fffffff2bb0) at /home/henri/Code/nanobind/include/nanobind/eigen/sparse.h:107
        src = @0x7fffffff29c8: {
          <Eigen::SparseCompressedBase<Eigen::SparseMatrix<float, 1, int> >> = {
            <Eigen::SparseMatrixBase<Eigen::SparseMatrix<float, 1, int> >> = {
              <Eigen::EigenBase<Eigen::SparseMatrix<float, 1, int> >> = {<No data fields>}, 
              members of Eigen::SparseMatrixBase<Eigen::SparseMatrix<float, 1, int> >:
              m_isRValue = false
            }, <No data fields>}, 
          members of Eigen::SparseMatrix<float, 1, int>:
          m_outerSize = 5,
          m_innerSize = 6,
          m_outerIndex = 0xabc890,
          m_innerNonZeros = 0x0,
          m_data = {
            m_values = 0xac7f30,
            m_indices = 0xacd210,
            m_size = 13,
            m_allocatedSize = 13
          }
        }
        matrix_type = {
          <nanobind::handle> = {
            <nanobind::detail::api<nanobind::handle>> = {
              <nanobind::detail::api_tag> = {<No data fields>}, <No data fields>}, 
            members of nanobind::handle:
            static Name = <optimized out>,
            m_ptr = 0xae74c0
          }, 
          members of nanobind::object:
          static Name = <optimized out>
        }
        data_shape = {13}
        outerIndices_shape = {6}
        data = {
          m_handle = 0x7fffae42b970,
          m_tensor = {
            data = 0xac7f30,
            device = {
              device_type = 1,
              device_id = 0
            },
            ndim = 1,
            dtype = {
              code = 2 '\002',
              bits = 32 ' ',
              lanes = 1
            },
            shape = 0x7fff6ba88150,
            strides = 0x7fff6ba88120,
            byte_offset = 0
          }
        }
        outerIndices = {
          m_handle = 0x7fffae42b790,
          m_tensor = {
            data = 0xabc890,
            device = {
              device_type = 1,
              device_id = 0
            },
            ndim = 1,
            dtype = {
              code = 0 '\000',
              bits = 32 ' ',
              lanes = 1
            },
            shape = 0x7fff6ba88140,
            strides = 0x7fff6ba88160,
            byte_offset = 0
          }
        }
        innerIndices = {
          m_handle = 0x7fffae42b870,
          m_tensor = {
            data = 0xacd210,
            device = {
              device_type = 1,
              device_id = 0
            },
            ndim = 1,
            dtype = {
              code = 0 '\000',
              bits = 32 ' ',
              lanes = 1
            },
            shape = 0x7fff6ba88170,
            strides = 0x7fff6ba88180,
            byte_offset = 0
          }
        }
#8  0x00007fffae4b31be in nanobind::detail::type_caster<Eigen::SparseMatrix<float, 1, int>, int>::from_cpp (v=..., policy=nanobind::rv_policy::move, cleanup=0x7fffffff2bb0) at /home/henri/Code/nanobind/include/nanobind/eigen/sparse.h:99
No locals.
#9  0x00007fffae494d9f in nanobind::detail::func_create<false, true, nanobind_init_test_eigen_ext(nanobind::module_&)::$_19, Eigen::SparseMatrix<float, 1, int>, , , nanobind::scope, nanobind::name>(nanobind_init_test_eigen_ext(nanobind::module_&)::$_19&&, Eigen::SparseMatrix<float, 1, int> (*)(), std::integer_sequence<unsigned long>, nanobind::scope const&, nanobind::name const&)::{lambda(void*, _object**, unsigned char*, nanobind::rv_policy, nanobind::detail::cleanup_list*)#1}::operator()(void*, _object**, unsigned char*, nanobind::rv_policy, nanobind::detail::cleanup_list*) const (this=0x0, p=0x7fffaee26468, args=0x7fffaee72538, args_flags=0x7fffffff2bf0 "\001\001\001\001\001\001\001\001", policy=nanobind::rv_policy::automatic, cleanup=0x7fffffff2bb0) at /home/henri/Code/nanobind/include/nanobind/nb_func.h:128
        cap = 0x7fffaee26468
        result = 0x7ffff660d3b0
#10 nanobind::detail::func_create<false, true, nanobind_init_test_eigen_ext(nanobind::module_&)::$_19, Eigen::SparseMatrix<float, 1, int>, , , nanobind::scope, nanobind::name>(nanobind_init_test_eigen_ext(nanobind::module_&)::$_19&&, Eigen::SparseMatrix<float, 1, int> (*)(), std::integer_sequence<unsigned long>, nanobind::scope const&, nanobind::name const&)::{lambda(void*, _object**, unsigned char*, nanobind::rv_policy, nanobind::detail::cleanup_list*)#1}::__invoke(void*, _object**, unsigned char*, nanobind::rv_policy, nanobind::detail::cleanup_list*) (p=0x7fffaee26468, args=0x7fffaee72538, args_flags=0x7fffffff2bf0 "\001\001\001\001\001\001\001\001", policy=nanobind::rv_policy::automatic, cleanup=0x7fffffff2bb0) at /home/henri/Code/nanobind/include/nanobind/nb_func.h:105
No locals.
#11 0x00007fffae4c4f3f in nanobind::detail::nb_func_vectorcall_simple (self=0x7fffaee26440, args_in=0x7fffaee72538, nargsf=9223372036854775808, kwargs_in=0x0) at /home/henri/Code/nanobind/src/nb_func.cpp:732
        f = 0x7fffaee26468
        e = @0x40d0f0: {
          <std::exception> = {<No data fields>}, 
          members of nanobind::python_error:
          m_type = 0x0,
          m_value = 0x439fb0,
          m_trace = 0x7fffaee723b0,
          m_what = 0x28 <error: Cannot access memory at address 0x28>
        }
        k = 0
        pass = 1
        args_flags = "\001\001\001\001\001\001\001\001"
        fr = 0x7fffaee26468
        count = 1
        nargs_in = 0
        is_method = false
        is_constructor = false
        self_flags = 0
        result = 0x0
        self_arg = 0x0
        cleanup = {
          static Small = 6,
          m_size = 1,
          m_capacity = 6,
          m_data = 0x7fffffff2bc0,
          m_local = {0x0, 0xac2aa7be639ae800, 0x40d0f0, 0x7fffaee723b0, 0x40d0f0, 0x7ffff66483b0}
        }
        error_handler = 0x0
        fail = false
        none_ptr = 0x7ffff7fafe30 <_Py_NoneStruct>
#12 0x00007ffff7c49cf2 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#13 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#14 0x00007ffff7c41169 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#15 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#16 0x00007ffff7c41169 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#17 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#18 0x00007ffff7c49cf2 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#19 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#20 0x00007ffff7cf7cd8 in method_vectorcall () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#21 0x00007ffff7c49cf2 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#22 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#23 0x00007ffff7d02736 in _PyObject_Call_Prepend () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#24 0x00007ffff7d43b49 in slot_tp_call () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#25 0x00007ffff7c95c84 in _PyObject_MakeTpCall () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#26 0x00007ffff7c47f69 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#27 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#28 0x00007ffff7c475b9 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#29 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#30 0x00007ffff7c41169 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#31 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#32 0x00007ffff7c49cf2 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#33 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#34 0x00007ffff7cf7cd8 in method_vectorcall () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#35 0x00007ffff7c49cf2 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#36 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#37 0x00007ffff7d02736 in _PyObject_Call_Prepend () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#38 0x00007ffff7d43b49 in slot_tp_call () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#39 0x00007ffff7ca7b39 in PyObject_Call () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#40 0x00007ffff7c41169 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#41 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#42 0x00007ffff7c4669f in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#43 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#44 0x00007ffff7cf7cd8 in method_vectorcall () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#45 0x00007ffff7c46344 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#46 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#47 0x00007ffff7c41169 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#48 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#49 0x00007ffff7c4669f in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#50 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#51 0x00007ffff7c46344 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#52 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#53 0x00007ffff7c41169 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#54 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#55 0x00007ffff7c49cf2 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#56 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#57 0x00007ffff7cf7cd8 in method_vectorcall () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#58 0x00007ffff7c49cf2 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#59 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#60 0x00007ffff7d02736 in _PyObject_Call_Prepend () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#61 0x00007ffff7d43b49 in slot_tp_call () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#62 0x00007ffff7c95c84 in _PyObject_MakeTpCall () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#63 0x00007ffff7c47f69 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#64 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#65 0x00007ffff7c41169 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#66 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#67 0x00007ffff7c49cf2 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#68 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#69 0x00007ffff7cf7cd8 in method_vectorcall () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#70 0x00007ffff7c49cf2 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#71 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#72 0x00007ffff7d02736 in _PyObject_Call_Prepend () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#73 0x00007ffff7d43b49 in slot_tp_call () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#74 0x00007ffff7c95c84 in _PyObject_MakeTpCall () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#75 0x00007ffff7c47f69 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#76 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#77 0x00007ffff7c4669f in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#78 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#79 0x00007ffff7c4669f in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#80 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#81 0x00007ffff7c41169 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#82 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#83 0x00007ffff7c49cf2 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#84 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#85 0x00007ffff7cf7cd8 in method_vectorcall () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#86 0x00007ffff7c49cf2 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#87 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#88 0x00007ffff7d02736 in _PyObject_Call_Prepend () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#89 0x00007ffff7d43b49 in slot_tp_call () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#90 0x00007ffff7c95c84 in _PyObject_MakeTpCall () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#91 0x00007ffff7c47f69 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#92 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#93 0x00007ffff7c4669f in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#94 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#95 0x00007ffff7c49cf2 in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#96 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#97 0x00007ffff7dc18e8 in PyEval_EvalCode () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#98 0x00007ffff7e590ed in builtin_exec () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#99 0x00007ffff7db69bf in cfunction_vectorcall_FASTCALL () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#100 0x00007ffff7c4669f in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#101 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#102 0x00007ffff7c4669f in _PyEval_EvalFrameDefault () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#103 0x00007ffff7dc127f in _PyEval_Vector () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#104 0x00007ffff7e4e5c7 in pymain_run_module () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#105 0x00007ffff7e59f0b in Py_RunMain () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#106 0x00007ffff7e5a535 in Py_BytesMain () from /nix/store/5axq6aw8j3vcs2m7gi440cwpcckl7ql9-python3-3.10.9/lib/libpython3.10.so.1.0
No symbol table info available.
#107 0x00007ffff78bd24e in __libc_start_call_main () from /nix/store/9xfad3b5z4y00mzmk2wnn4900q0qmxns-glibc-2.35-224/lib/libc.so.6
No symbol table info available.
#108 0x00007ffff78bd309 in __libc_start_main_impl () from /nix/store/9xfad3b5z4y00mzmk2wnn4900q0qmxns-glibc-2.35-224/lib/libc.so.6
No symbol table info available.
#109 0x0000000000401075 in _start ()
No symbol table info available.
```

</details>

hmenke replied:
Ah no, wait. It looks like `module_::import_("scipy.sparse")` throws and causes terminate.

wjakob commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,125 @@
+	auto obj = borrow<object>(src);
+	try {
+	    object matrix_type = module_::import_("scipy.sparse").attr(row_major ? "csr_matrix" : "csc_matrix");
+	    if (!obj.type().is(matrix_type)) {
```
can skip the extra braces here.

wjakob commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,125 @@
+
+    NB_TYPE_CASTER(T, const_name<row_major>("scipy.sparse.csr_matrix[",
+					    "scipy.sparse.csc_matrix[")
+		   + make_caster<Scalar>::Name + const_name("]"));
```
`make_caster<Scalar>` -> `ScalarCaster`

hmenke replied:
But `ScalarCaster = make_caster<ScalarTensor> ≠ make_caster<Scalar>`. I think we do want `make_caster<Scalar>::Name` here.

wjakob replied:
my bad.

wjakob commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,125 @@
+
+        value = EigenMapSparseMatrix<Scalar,
+                                     T::Flags &(Eigen::RowMajor | Eigen::ColMajor),
+                                     StorageIndex>(std::move(rows),
```
Let's remove the use of ``std::move`` to pass integer values here.

wjakob commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,125 @@
+	const size_t data_shape[] = { static_cast<size_t>(v.nonZeros()) };
+	const size_t outerIndices_shape[] = { static_cast<size_t>((row_major ? v.rows() : v.cols()) + 1) };
+        ScalarTensor data(src.valuePtr(), 1, data_shape);
+        StorageIndexTensor outerIndices(src.outerIndexPtr(), 1, outerIndices_shape);
```
`outerIndices` -> `outer_indices`.

wjakob reviewed (commented):
This looks great. Just a few minor extra comments.

wjakob commented:
It would still be good to figure out a way to make `nanobind/eigen/dense.h` not depend on `Eigen/SparseCore` (see my earlier comment). Following this, would you be able to squash the commits and rebase on `master`?

wjakob commented:
By the way, you may also need to change `.github/workflows/ci.yml` to install the relevant SciPy packages so that your changes are actually tested.

wjakob commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,124 @@
+/*
+    nanobind/eigen/dense.h: type casters for dense Eigen
```
update

wjakob commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,124 @@
+    nanobind/eigen/dense.h: type casters for dense Eigen
+    vectors and matrices
+
+    Copyright (c) 2023 Wenzel Jakob
```
Suggested: ``Copyright (c) 2023 Henri Menke and Wenzel Jakob``

wjakob commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,124 @@
+template <typename Scalar, int Flags, typename StorageIndex>
+using EigenMapSparseMatrix = Eigen::Map<Eigen::SparseMatrix<Scalar, Flags, StorageIndex>>;
+
+/// Caster for Eigen::SparseMatrix<T>
```
Is it really just ``Eigen::SparseMatrix<T>`` or also other types?

wjakob commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,124 @@
+    StorageIndexCaster storageindex_caster;
+
+    bool from_python(handle src, uint8_t flags, cleanup_list *cleanup) noexcept {
+        auto obj = borrow<object>(src);
```
The `<object>` template argument is not needed. Replace `auto` with `object`.

wjakob commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,124 @@
+            return false;
+        }
+
+        value = EigenMapSparseMatrix<Scalar,
```
The formatting is a bit unwieldy here. Perhaps you can define this type with a `using SparseMap = ..` declaration at the top of the caster and then reference it here?

hmenke commented:
Rebased and squashed. Checks are enabled, but as already noted something throws in `from_cpp`, apparently during `make_tuple` in the return statement.

[Backtrace](https://github.com/wjakob/nanobind/files/10743367/gdb.txt)

wjakob commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    static handle from_cpp(const T &v, rv_policy policy, cleanup_list *cleanup) noexcept {
+        // FIXME: Looks like UB. Is there anything that can be done about this?
```
This indeed doesn't seem like a good design. I suggest the following instead:

Option 1: give up (untested)

```cpp
if (!v.isCompressed()) {
    PyErr_SetString(PyExc_ValueError, "nanobind: unable to return an Eigen sparse matrix that is not in a compressed format. Please call `.makeCompressed()` before returning the value on the C++ end.");
    return handle();
}
```

Option 2: copy to turn the matrix into a compressed format (untested):

```cpp
std::unique_ptr<T> temp;
T *src = &v;

if (!v.isCompressed()) {
    temp = make_unique<T>(v);
    temp->makeCompressed();
    src = temp.get();
}

```

hmenke replied:
I chose option 1. Time will tell whether that was a good decision.

wjakob commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,124 @@
+        StorageIndexTensor outer_indices(src.outerIndexPtr(), 1, outerIndices_shape);
+        StorageIndexTensor inner_indices(src.innerIndexPtr(), 1, data_shape);
+
+        return matrix_type(make_tuple(
```
This function call can fail with an exception, and that is not allowed by a caster. Once more, I think there is a whole bunch of code that must be wrapped:

```cpp
try {
    // code that may throw
} catch (const python_error &e) {
    e.restore();
    return handle();
}
```

wjakob commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,124 @@
+    using Index = typename T::Index;
+    static constexpr bool row_major = T::IsRowMajor;
+
+    using ScalarTensor = tensor<Scalar, shape<any>, c_contig, device::cpu>;
```
I think one crucial thing is still missing in these type annotations: `numpy`! Otherwise, nanobind will call the SciPy sparse matrix constructor with a bunch of DLPack capsules, and SciPy does not know what to do with hose.

hmenke replied:
Thanks! That's it.

wjakob commented:
Please also add a changelog entry ;)

wjakob commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,115 @@
+#include <nanobind/eigen/dense.h>
+#include <Eigen/SparseCore>
+
+static_assert(EIGEN_VERSION_AT_LEAST(3, 3, 0),
```
I don't think you need to repeat the version check, if it is already in `eigen/dense.h`.

wjakob commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,112 @@
+/*
+    nanobind/eigen/sparse.h: type casters for sparse Eigen
+    vectors and matrices
```
We do not support sparse vectors though, right? I remember there being a pybind11 bug report about it.

hmenke commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,118 @@
+			    "Please call `.makeCompressed()` before returning the value on the C++ end.");
+	    return handle();
+	}
+        T & src = const_cast<T &>(v);
+
+        object matrix_type = module_::import_("scipy.sparse").attr(row_major ? "csr_matrix" : "csc_matrix");
+
+        const size_t data_shape[] = { (size_t)v.nonZeros() };
+        const size_t outer_indices_shape[] = { (size_t)((row_major ? v.rows() : v.cols()) + 1) };
+        ScalarTensor data(src.valuePtr(), 1, data_shape);
```
`const_cast` is still needed because `nb::tensor` has no constructor for `const void *`. Is there a deeper reason for that?

wjakob replied:
we currently don't have a way of making read-only tensors, though that is a feature I would like to support in the future.

hmenke commented:
I have to double check the minimum Eigen version. The current 3.3.0 seems too low (it's also from way back in 2016, i.e. before C++17). In principle I would also like to improve the test coverage a bit. Right now it's limited to the pybind11 tests.

wjakob commented:
Agreed, the tests seem a bit, ahem -- sparse. ;)

wjakob commented:
Another aspect I realized today is that one feature in the dense caster is not shared by this sparse version. In the `from_cpp` part, when the return value policy is `move`, nanobind can perform a zero-copy transfer by move-constructing the Eigen object into a new heap-allocated instance, and then tying the lifetime of the newly created NumPy wrapper using a `capsule` owner object.

See the special case here: https://github.com/wjakob/nanobind/blob/master/include/nanobind/eigen/dense.h#L132

hmenke commented on include/nanobind/eigen/sparse.h:
```diff
@@ -109,16 +109,26 @@ template <typename T> struct type_caster<T, enable_if_t<is_eigen_sparse_v<T>>> {
-        StorageIndexTensor inner_indices(src.innerIndexPtr(), 1, data_shape);
+        const size_t outer_indices_shape[] = { (size_t)((row_major ? rows : cols) + 1) };
+
+        T *src = std::addressof(const_cast<T &>(v));
+        object owner;
+        if (policy == rv_policy::move) {
+            src = new T(std::move(v));
+            owner = capsule(src, [](void *p) noexcept { delete (T *) p; });
+        }
+
+        ScalarTensor data(src->valuePtr(), 1, data_shape, owner);
+        StorageIndexTensor outer_indices(src->outerIndexPtr(), 1, outer_indices_shape, owner);
+        StorageIndexTensor inner_indices(src->innerIndexPtr(), 1, data_shape, owner);
```
Do I need one owner per tensor or is this okay? I ran the existing tests with ASan on and there were no failure, but I'm also not sure whether I'm triggering this optimization right now.

wjakob replied:
A single owner shared by all tensors is good. Each tensor will reference the Python capsule, and it can only be deleted when they all expire.

wjakob commented:
Any further changes planned?

hmenke commented:
I would like to test returning `Eigen::Ref<Eigen::SparseMatrix>` and  `Eigen::Map<Eigen::SparseMatrix>` from C++. This could also go in a later PR if you prefer to get the feature in now. I'll quickly rebase.

wjakob commented:
I think that it would be great to have `Eigen::Ref<Eigen::SparseMatrix>` and `Eigen::Map<Eigen::SparseMatrix>`. You don't need to write documentation, I already started myself on writing some for Eigen/Dense and that can easily be adapted/extended.

Note that there was a change on `master` that impacts this PR: the `nb::tensor<..>` type was renamed to `nb::ndarray<..>`

hmenke commented:
I've tried to get `Map` and `Ref` support up and running but failed so far. For now I have added stubs for these implementation so that if anybody attempts to use this, the compiler errors are a little bit less useless.

wjakob commented:
Okay, fair enough. FYI, I added preliminary documentation for the dense part in the meantime: https://nanobind.readthedocs.io/en/latest/eigen.html — can you address the remaining conflicts so that the PR can be merged? (should just be a matter of moving the sparse-specific text into a subsection).

hmenke commented:
Done. I think it'd be good to run the testsuite again before merging just to be sure.

wjakob commented:
Awesome, all passes! I'll squash these into a single commit. Thank you for all of this hard work!

hmenke commented on include/nanobind/eigen/sparse.h:
```diff
@@ -0,0 +1,177 @@
+        StorageIndexNDArray inner_indices(src->innerIndexPtr(), 1, data_shape, owner);
+
+        try {
+            return matrix_type(make_tuple(
+                                   std::move(data), std::move(inner_indices), std::move(outer_indices)),
+                               make_tuple(rows, cols))
```
I've noticed that this sometimes create ADL problems, because `nanobind::make_tuple` competes with `std::make_tuple`. A simple fix would be to fully qualify `make_tuple`, but is there a better solution? @wjakob 
```suggestion
            return matrix_type(nanobind::make_tuple(
                                   std::move(data), std::move(inner_indices), std::move(outer_indices)),
                               make_tuple(rows, cols))
```
